### ISSUE TITLE
Support for Voltage Sensor

### ISSUE THREAD

**Issue by BoLtWD** [RotorHazard/RotorHazard#61]
I dont know if this is possible but was wondering if a Voltage Sensor could be used to monitor the battery and display the volts with possible display warning when battery is at a users predesignated low level.  

Voltage Sensor
Https://www.dhgate.com/product/voltage-sensor-for-arduino-dc-raspberry-pi/401118628.html

Analogue-to-Digital converter
https://projects.raspberrypi.org/en/projects/physical-computing/15

**Comment by BoLtWD** [comment]
Ok think I have seen a problem, as the MCP3008 Analogue to Digital Converter uses SPI pins GPIO 8-10 you could not use the LED that are normally on GPIO 10. I guess this approach or idea is a no go..

**Comment by Ryanf55** [comment]
This is a pretty neat way to do it. 
https://www.raspberrypi.org/forums/viewtopic.php?f=29&t=21287&p=447537#p447537

**Comment by jrwrodgers** [comment]
@BoLtWD I was thinking of something similar when building my system. I think its a good idea to have a input into the pi logging the voltage of the power supply and then at the top of webpage battery percentage bar. 

I have mocked up some logging on my system using an MCP3008 and an SPI connection with no conflict of pins for the LED.

![image](https://user-images.githubusercontent.com/6489861/54159808-cfb9c000-4445-11e9-9039-c88db1d1d220.png)

MCP3008 VDD to Raspberry Pi 3.3V
MCP3008 VREF to Raspberry Pi 3.3V 
MCP3008 AGND to Raspberry Pi GND
MCP3008 DGND to Raspberry Pi GND
MCP3008 CLK to Raspberry Pi SCLK
MCP3008 DOUT to Raspberry Pi MISO
MCP3008 DIN to Raspberry Pi MOSI
MCP3008 CS/SHDN to Raspberry Pi CE0

CH0 is then connected to my 4S LiPO via a voltage divider:

![image](https://user-images.githubusercontent.com/6489861/54159925-190a0f80-4446-11e9-95ab-48439fbd1d16.png)

When fully charged the voltage is:
Vmax = 4.7/(4.7+22) * 16.8V = 2.96V or 918 (digital MCP output)

When low the volatge is:
Vmin = 4.7/(4.7+22)*15.4V = 2.76V or 856 (digital MCP output)

I can poll the MCP3008 using a simple python code to return the voltage as an integer.  
If I branch and add this code and functionality but would anyone be able to help out with the server webpage integration?

**Comment by BoLtWD** [comment]
This looks fantastic and can see how this would finally bring a user interface to the timer that can help a race director determine a life expectancy of a battery so not to have problums during a race.

There may need to be options to select different power in options in the user interface for 12v PSU, 3s, 4s possibly by setting the Min, Max and ALARM values from within the user interface to suit a specific criteria. By doing this it would probably need these values saved to the database.

I wish I could code to help you out but this is where my abilities in coding are lower than basic.

**Comment by Ryanf55** [comment]
I am happy to add code to support a voltage alarm. Were you thinking of adding like a buzzer to the box, a display, or some LED's? Or is this for the web interface?

If it's arduino, I can do it.

**Comment by BoLtWD** [comment]
The principal I was thinking of was to bring an indicator to the race director using the Web interface so that the race director can make an informed choice to start a race knowing that there should be battery to complete a race before the battery running out.

 

I find the best checker is a basic LIPO checker with alarms connected to the balance leads but this is no good if you want to keep an eye on your battery and your timer is suck in the middle of a field under a box. This is why I would like to see this to help a race director who is nowhere near the timer.

 

The flyers I fly with show protest if the LIPO checkers start beeping during a race as this upsets there flying, so they all under protest land there Quads on or around the timer when the LIPO checkers starts beeping and indicating low battery.

 

I was thinking something like a simple battery indicator with an option to set a warning of some kind at a low value.

**Comment by HazardCreative** [comment]
To move forward with including this, I'd really like to see a PCB design with a section that can be optionally populated with the components for it.

Doing it like this would create an easily accessible "standard" for how the Pi gets its voltage data. Without a standard, each implementation that people find and use may have different requirements and the data they output may look different and need processed differently. I don't want to get into supporting many different methods for this, but I'd be happy to support one that's easy for any user to add in if they desire. The popularity of the D5's PCB suggests that for many, building a circuit from a diagram is outside their comfort zone but populating a PCB is more attainable. Of course the circuit design will still be published so those that want to do something differently are able to do so.

I'd also like it to have as little hardware requirement as possible. The polling rate for a voltage sensor can be very low (even 1Hz may be more than necessary) and there's no need for it to have priority. To use an existing bus instead of more unique resources on the Pi would be better for allowing future enhancements that could be more time-critical. Is there any way to put this on the existing SPI bus alongside the LED strip? Or is there another method that would be better? My background is not in hardware so I don't feel experienced enough to say if this is the best solution. At the least, I'd like to hear from @ethomas997 on this.

Once the data is in Python, I'm completely confident I can make a nice display and warning system for the web UI.

I don't have any experience in PCB design, so we'll need a partner with that capability. There are other PCB revisions I'd like to have made as well.

**Comment by jrwrodgers** [comment]
I think simplest way would be using one of the nano analogue channels with a voltage divider. We can then poll the voltage at whatever frequency we like in the node.ino code and convert to a battery voltage with some defined constants. I'll test it out and report back. This minimizes any hardware changes and uses the existing bus.

**Comment by pulquero** [comment]
I'm ordering some INA219 I2C sensors to try something similar with those. They're cheap and measure both voltage and current. I'm gonna put one on the battery and another on the 5V to the Pi.

**Comment by Ryanf55** [comment]
> I'm ordering some INA219 I2C sensors to try something similar with those. They're cheap and measure both voltage and current. I'm gonna put one on the battery and another on the 5V to the Pi.

Not bad. I ordered a ADS1115. It's the same price, but has 4 voltage channels. Your hardware looks easier. I also wrote some arduino like jrwrodgers for sensing voltage by adding a new I2C command, but have yet to test. 

I was under the impression you configure in the GUI the voltage divider ratio and which node number has voltage sensing soldered up. Then, only that arduino gets referenced for a voltage request, and all the Arduino's have the same code.
